File: fsharp/data_structures/linked_list.fs
(*
Linked list data structure example
Copyright 2016, Sjors van Gelderen
*)

module Program

// Linked list data structure
type LinkedList<'a> =
    | Node of 'a * LinkedList<'a>
    | Empty

(*
Insert an element into the list
Complexity: O(1)
*)
let rec insert element list =
    printfn "Inserting element %A" element
    Node (element, list)
        
(*
Delete an element from the list
Complexity: O(n)
*)
let rec delete element list =
    match list with
    | Node (value, tail) ->
        if value = element then
            printfn "Deleting element %A" element
            tail
        else
            Node (value, delete element tail)
    | Empty ->
        printfn "Element %A not in collection" element
        list

(*
Find an element in the list
Complexity: O(n)
*)
let rec search element list =
    match list with
    | Node (value, tail) ->
        if value = element then
            printfn "Found element %A" element
        else
            search element tail
    | Empty ->
        printfn "Element %A not in list" element

(*
Print all values in the list
Complexity: O(n)
*)
let rec print list =
    match list with
    | Node (value, tail) ->
        printfn "%A" value
        print tail
    | Empty ->
        ()

[<EntryPoint>]
let main args =
    printfn "Linked list data structure example - \
             Copyright 2016, Sjors van Gelderen"
    
    let list = Node(10, Node(12, Node(16, Empty)))
    print list
    
    let list = Empty |> insert 10 |> insert 12 |> insert 50
    print list

    let list = list |> delete 50 |> delete 24 |> delete 5
    print list

    List.iter (fun elem -> search elem list) [ 10; 2; 56; 12 ]
    
    0

File: python_3/data_structures/linked_list.py
"""Linked list data structure
Copyright 2016, Sjors van Gelderen
"""

# Empty list
class Empty:
    def __init__(self):
        self.is_empty = True

    
# List segment
class Segment:
    def __init__(self, _value, _tail):
        self.value = _value
        self.tail = _tail
        self.is_empty = False


# Full list functionality
class List:
    def __init__(self):
        self.segment = Empty()
    
    def __str__(self):
        values = []
        segment = self.segment
        while not segment.is_empty:
            values.append(str(segment.value))
            segment = segment.tail
        
        return " ".join(values)
    
    def populate(self, *_values):
        for value in _values:
            self.segment = Segment(value, self.segment)

    def reverse(self):
        result = List()
        values = []
        segment = self.segment
        
        while not segment.is_empty:
            values.append(segment.value)
            segment = segment.tail

        for value in values:
            result.segment = Segment(value, result.segment)

        self.segment = result.segment

            
# Main program logic
def program():
    linked_list = List()
    linked_list.populate(1, 2, 3, 4)
    print(linked_list)
    linked_list.reverse()
    print(linked_list)
            
program()
